### PR TITLE
WIP: Configuration of required attributes for registration

### LINKED ISSUE #67: Configuration of required attributes for registration
At this time one can only configure what fields show up on the registration page. Which fields are required or not is statically implemented and so cannot be changed by a deployer/operator. I think we should make it possible for a deployer/operator to define which fields are required or not. @dacrome already made up an example how to specify that in the properties file: https://gist.github.com/dacrome/31edbb3b9e894258e937

### PR DESCRIPTION
**Pull request by dacrome**
Resolves #67

### CODE REVIEW

**Comment by dacrome** [comment]
Corresponding ITs: https://github.com/osiam/connector4java-integration-tests/pull/304

File: docs/migration.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+In `htmlfield.html` you need to change all `required` occurrences with
+`requiredFieldMapping.get(attribute)`.
+
+Before:
+
+    th:required="${required}"
+    
+After:
+
+    th:required="${requiredFieldMapping.get(attribute)}"
```
Comment (tkrille) [inline docs/migration.md]: I think, in Thymeleaf you can do `requiredFieldMapping[attribute]`. Reads much better.

**Comment by tkrille** [comment]
You replaced `allowedFields`, which held all fields that a user can enter on the registration page, with `requiredFieldMapping`, which holds a mapping from an allowed field name to it's required state (`true`/`false`). This seems to me like a data structure from hell. The name is also misleading: It appears  at first, that it only contains the required fields. Here's a suggestion: make it a `List<Field>` with `Field` (Please find a better name) containing the name and whether it's required or not. I mean, you already changed the whole thing, why not create a DTOish class to get the data into the template?

**Comment by tkrille** [comment]
Why should `htmlfield.html` and `extensionInput.html` bother themselves with getting the required state? I would not change this. They are like helper functions, that just get input and return output without other collaborations beyond that.

**Comment by tkrille** [comment]
We should use `th:each` in the default registration template and somehow get the field type by the attribute's name. All available attribute names with their corresponding field type are already there in the template. Might be better to have them in the code.

**Comment by dacrome** [comment]
Thanks @tkrille for your review. You are absolutely right, I need to revise this pull request.

I think I changed it to the following:

* `Field` bean class, which holds the `name`, `required`, `type`, `isExtension`. This is also the structure which is configurable: `name:required:type`
* In the code I make a pre check if the field name is available, if not I log a warning and ignore the field
* In the template I iterate over the field list with `th:each` and use the existing helper functions of `htmlfield.html`

**Comment by dacrome** [comment]
@tkrille had the good idea to use yaml to configure the fields and extensions. So I will first switch to yaml (#134) before I finish this pull request.

**Comment by dacrome** [comment]
What is missing: The fields have to be also checked on server side, so a dynamic bean validation is needed to check which fields are configured as required. But this easy and [already done for other constraints](https://github.com/osiam/addon-self-administration/blob/de8228c97e07c2d1668d0ba12d62d5f41c97196f/src/main/java/org/osiam/addons/self_administration/validators/EqualPasswordValidator.java).

File: src/main/java/org/osiam/addons/self_administration/Config.java [hunk 2]
```diff
@@ -118,43 +131,114 @@
...
-    private String[] allAllowedFields;
+    private HtmlField[] allAllowedFields;
+    private String[] allowedFieldNames;
 
     private boolean confirmPasswordRequired = true;
 
     @Autowired
-    private void createAllAllowedFields(@Value("${org.osiam.html.form.fields:}") String[] allowedFields,
-            @Value("${org.osiam.html.form.extensions:}") String[] extensions) {
-        List<String> collectedFields = new ArrayList<>();
```
Comment (timopick) [inline src/main/java/org/osiam/addons/self_administration/Config.java]: Codacy Issue found: [Avoid really long methods.](https://www.codacy.com/app/t-pick/addon-self-administration/file/2113712806/issues/source?bid=3145222&fileBranchId=3145225#l140)

File: src/main/java/org/osiam/addons/self_administration/Config.java [hunk 3]
```diff
@@ -118,43 +131,114 @@
...
+                // urn:required
+                if (type.equals("true") || type.equals("false")) {
+                    required = Boolean.valueOf(type);
+                    htmlField.setRequired(required);
+                    htmlField.setName(trimmedField.substring(0, lastIndexOf));
+                } else {
+                    // urn:required:type
+                    final String urnRequired = trimmedField.substring(0, lastIndexOf - 1);
+                    lastIndexOf = urnRequired.lastIndexOf(":");
+                    final String requiredString = trimmedField.substring(lastIndexOf + 1, urnRequired.length());
```
Comment (timopick) [inline src/main/java/org/osiam/addons/self_administration/Config.java]: Codacy Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/t-pick/addon-self-administration/file/2113712806/issues/source?bid=3145222&fileBranchId=3145225#l224)

File: src/main/java/org/osiam/addons/self_administration/Config.java (hunk 2 above)
Comment (timopick) [inline src/main/java/org/osiam/addons/self_administration/Config.java]: Codacy Issue found: [The method createAllAllowedFields() has an NPath complexity of 5280](https://www.codacy.com/app/t-pick/addon-self-administration/file/2113712806/issues/source?bid=3145222&fileBranchId=3145225#l140)

File: src/main/java/org/osiam/addons/self_administration/Config.java [hunk 4]
```diff
@@ -118,43 +131,114 @@
...
+                int lastIndexOf = trimmedField.lastIndexOf(":");
+                if (lastIndexOf == -1) {
+                    throw new ConfigurationException("The extension configuration must be in the form field:required " +
+                            "e.g. 'extensions['urn:client:extension'].fields['age']:true' or just the field e.g. " +
+                            "'extensions['urn:client:extension'].fields['age']', but it is: " + trimmedField);
+                }
+
+                final String type = trimmedField.substring(lastIndexOf + 1, trimmedField.length());
+                final boolean required;
+                // urn:required
```
Comment (timopick) [inline src/main/java/org/osiam/addons/self_administration/Config.java]: Codacy Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/t-pick/addon-self-administration/file/2113712806/issues/source?bid=3145222&fileBranchId=3145225#l215)

File: src/main/java/org/osiam/addons/self_administration/Config.java (hunk 2 above)
Comment (timopick) [inline src/main/java/org/osiam/addons/self_administration/Config.java]: Codacy Issue found: [Avoid really long methods.](https://www.codacy.com/app/t-pick/addon-self-administration/file/2113712806/issues/source?bid=3145222&fileBranchId=3145225#l140)

File: src/main/java/org/osiam/addons/self_administration/Config.java (hunk 3 above)
Comment (timopick) [inline src/main/java/org/osiam/addons/self_administration/Config.java]: Codacy Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/t-pick/addon-self-administration/file/2113712806/issues/source?bid=3145222&fileBranchId=3145225#l224)

File: src/main/java/org/osiam/addons/self_administration/Config.java (hunk 2 above)
Comment (timopick) [inline src/main/java/org/osiam/addons/self_administration/Config.java]: Codacy Issue found: [The method createAllAllowedFields() has an NPath complexity of 5280](https://www.codacy.com/app/t-pick/addon-self-administration/file/2113712806/issues/source?bid=3145222&fileBranchId=3145225#l140)

File: src/main/java/org/osiam/addons/self_administration/Config.java (hunk 4 above)
Comment (timopick) [inline src/main/java/org/osiam/addons/self_administration/Config.java]: Codacy Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/t-pick/addon-self-administration/file/2113712806/issues/source?bid=3145222&fileBranchId=3145225#l215)

**Comment by dacrome** [comment]
Beside the fact that this PR needs a whole rework, this change should be made after the Registration feature is moved into OSIAM and the configuration of the required fields can be created with yaml.

**Pull request closed without merging**